stetho/src/main/java/com/facebook/stetho/inspector/network/DownloadingAsyncPrettyPrinterFactory.java

```diff
@@ -30,53 +30,77 @@ public AsyncPrettyPrinter getInstance(final String headerName, final String head
       return null;
     }
     String uri = result.getSchemaUri();
-    URL schemaURL = parseURL(uri);
+    final URL schemaURL = parseURL(uri);
     if (schemaURL == null) {
-      return getErrorAsyncPrettyPrinter(headerName, headerValue);
+      String errorMessage = "[Failed to parse header: "
+          + headerName + " : " + headerValue + " ]";
+      return getErrorAsyncPrettyPrinter(errorMessage);
     } else {
-      ExecutorService  executorService = AsyncPrettyPrinterExecutorHolder.getExecutorService();
-      if (executorService == null) {
-        //last peer is unregistered...
-        return null;
-      }
-      final Future<String> response = executorService.submit(new Request(schemaURL));
-      return new AsyncPrettyPrinter() {
-        public void printTo(PrintWriter output, InputStream payload)
-            throws IOException {
-          try {
-            String schema;
+      final String schemaCache;
+      AsyncPrettyPrinterSchemaManager schemaFileManager
```

Comment:
long style note:

You should only create the `AsyncPrettyPrinter` once. Imagine you want to add a prefix in `doPrint`. Now you have to know that you need to modify it in 2 places. Prefer a paradigm like the following

```
String schema = AsyncPretyPrinterSchemaManager.getInstance().get(schemaUrl);
ExecutorService  executorService = AsyncPrettyPrinterExecutorHolder.getExecutorService();
if (schema == null && executorService != null) {
  final Future<String> response = executorService.submit(new Request(schemaURL));
  try {
    schema = response.get();
  } catch (...) { ... }
}
if (schema != null) {
  return new AsyncPrettyPrinter() {...};
} else {
  return null;
}
```

This really drops your line count down, makes it more readable, and leaves less room for bugs. The other option is to keep a `String errorMessage` object as null and send that in the  `doErrorPrint` like you are doing already.